src/shared/fdset.c

```diff
@@ -12,21 +12,21 @@
 #include "fdset.h"
 #include "log.h"
 #include "macro.h"
+#include "ordered-set.h"
 #include "parse-util.h"
 #include "path-util.h"
-#include "set.h"
 #include "stat-util.h"
 
-#define MAKE_SET(s) ((Set*) s)
+#define MAKE_SET(s) ((OrderedSet*) s)
```

Comment:
i am not convinced of using an OrderedSet for this btw, the requirement to have the fds sorted is too much of a hack for my taste.

What about this: switch this to be a Hashmap, and use the fact that even on 32bit the pointer space is at least twice as large as the fd space (because fds are positive signed, and pointers are unsigned)

then, provide two distinct put calls, one where we just an fd, i.e. the status quo ante. it just inserts the thing with key *and* value as INT_TO_PTR(fd+1). And then add another one, where caller specifies an fd and an index. For that you create two entries, on from with the key as index nr pointing to the fd nr, and one with fd nr pointing to the index nr. now the trick to distinguish the two cases is to make indexes negative.

or in other words:

```
fdset_put(fdset, fd) → hashmap_put(fdset, FD_TO_PTR(fd), FD_TO_PTR(fd))
```

and

```
fdset_put_with_index(fdset, fd, idx) → hashmap_put(fdset, FD_TO_PTR(fd), INT_TO_PTR(-idx-1))
                                     → hashmap_put(fdset, INT_TO_PTR(-idx-1), FD_TO_PTR(fd));
```

when you remove fds, then you have to delete two entries if the key and value of the entry you remove differ in sign.

net result: you have a single data structure where you can find everything *either* by index, *or* by fd number.

Reply:
When I put an fd in the set, I don't know what the index should be. Quite the opposite: I need the data structure to tell me what the index is. So it can't be an input here, so this scheme doesn't quite work.

Reply:
> The only way I can see this working is with externally accounting for the index, and carrying that around, which appears to me what you'd consider 'hacky', but I don't see another way so I'll go for that, if I am missing something please let me know

pushed as described, hashmap plus external index tracking